accepted/0000-hub-online-synchronization.md

```diff
@@ -0,0 +1,398 @@
+
+Configuration should be pushed from the HUB to peripherals, to avoid the need of direct management of the peripherals.
+
+For now this solution is focused on software channels only, but can be extended to synchronize more data types.
+
+
+## Define connection between HUB and Peripherals
+
+Each peripheral server can only have 1 Hub server (main). This will avoid dealing with problems like channel label conflicts between multiple main servers.
+Configure a hub server on a peripheral server will block access to a set of menu items like: "Admin" -> "Setup wizard" -> "Organization Credentials" and "Products" and "PAYG Connection" (similar to what we already have for ISSv1).
+We cannot add Hub connection if SCC credentials are defined, they should be mutually excluded.
+
+### Alternative 1:
+We can follow a similar approach to what exists on ISSv1. On the hub side we can define multiple peripheral servers to connect to by providing the FQDN and an authentication token.
+On the peripheral side we also need to define the Hub server FQDN and it would generate an Authentication token for he HUB.
+
+To define this connection the existing API namespace should be re-use (`sync.master` and `sync.slave`), with some adaptations to cope with the new requirements. Example of this is the need to define an authentication method between servers to the connection more secure.
+
+During development phase we should consider if we can join both namespaces in one name `hub`, for example.
+
+
+### Alternative 2:
+
+We need to establish a secure connection between Hub and Peripheral Server.
+For this we create a Peripheral configuration on the Hub server.
+It mainly defines the FQDN of the Peripheral and generate credentials to authenticate from the Hub.
+The Root CA must be configured as well when the peripheral server is already setup.
+
+This configuration can be exported and needs to be transferred to the Peripheral Server.
+This configuration include:
+   - the Root CA of the Hub
+   - the FQDN of the Hub
+   - the credentials to setup on the Peripheral Server
+
+The authentication credentials works only for a special server-to-server API and is not connected to any user on the hub or peripheral server.
+The new server-to-server API is required as the existing API is tied to the organization of the calling user. The new API should work organization independent.
+
+The Peripheral Server may already be configured. When restarted with this additional configuration, it configures the Hub configuration and setup the credentials
+that the Hub can connect and finish the configuration.
+
+We may want to provide the possibility to generate a full peripheral configuration which makes it possible to setup the peripheral server from scratch including the Hub configuration (similar to the proxy configuration).
+
+
+After the peripheral server is setup and the Hub can connect via the server-to-server API, the final configuration can be triggered by the Hub.
+This include:
```

Comment:
This can all be done in the same taskomatic process, with several multiple steps. Hub decide to apply configuration from HUB and then all this data is transferred and stored in the right locations.

Reply:
yes, but this happens typically every night. But when you setup a new peripheral, you want to trigger it "now".
So we need the button and a job which does it nightly.

Reply:
Sound fine to me. however, would be a option to have the nightly job and at the same time trigger a job execution for that peripheral when he is created or when an update to the configurations is made? This way we would avoid the button.

Reply:
Well, you first generate this config. This is a change and you trigger this update, but at that point in time the peripheral is not yet setup and the job would fail.
Than you configure the peripheral with this config. Maybe that's all what you want at that point in time, but you would like to know if the communication works now. You will ask for a button.
I think we need a button. Especially when an admin do not want to deploy all the changes directly to the peripheral server he is configuring on the hub side.

Reply:
ok, got you point. I forgot about this chicken & egg problem :)
It's better to have the button, worth case if users don't want they are not forced to click it.